api/src/auth/__tests__/auth.service.spec.ts

```diff
@@ -0,0 +1,301 @@
+        ...createAccountDto,
+        isVerified: false,
+        role: 'standard'
+      };
+      usersService.createUser.mockResolvedValue(mockUser);
+      const result = await authService.createAccount(createAccountDto);
+      expect(result).toMatchObject(mockUser);
+    });
+  });
+
+  describe('login', () => {
+    beforeEach(() => {
+      // can use spyOn insead: spyOn(usersService, 'findByEmail').mockResolvedValue(createUserDto);
+      usersService.findByEmail.mockResolvedValue(createUserDto);
+      cryptoService.comparePassword.mockResolvedValue(true);
+    });
+
+    it('should throw an UnauthorizedException when given an invalid email when calling usersService.findByEmail', () => {
+      const { password } = createUserDto;
+      usersService.findByEmail.mockResolvedValue(undefined);
+      expect(authService.login('invalid@example.com', password)).rejects.toBeInstanceOf(UnauthorizedException);
+    });
+
+    it('should throw an UnauthorizedException when given an invalid password when calling cryptoService.comparePassword', () => {
+      const { email } = createUserDto;
+      cryptoService.comparePassword.mockResolvedValue(false);
+      expect(authService.login(email, 'invalidPassword')).rejects.toBeInstanceOf(UnauthorizedException);
+    });
+
+    it('should return an access token when given valid credentials', async () => {
+      const { email, password } = createUserDto;
+      const result = await authService.login(email, password);
+      expect(result.accessToken).toEqual('accessToken');
+    });
+  });
+
+  describe('sendVerificationCode', () => {
+    const locale: Locale = 'en';
+
+    it('should throw NotFoundException if the user is not found when calling usersService.findByEmail', () => {
+      usersService.findByEmail.mockResolvedValue(undefined);
+      expect(authService.sendVerificationCode(currentUser, locale)).rejects.toBeInstanceOf(NotFoundException);
+    });
+
+    it('should use an existing verification code if the date is not expired', async () => {
+      const user: User = {
+        email: 'test@example.com',
+        firstName: 'Test',
+        hashedPassword: 'hashedPassword',
+        isVerified: true,
+        lastName: 'User',
+        role: 'standard',
+        verificationCode: {
+          attemptsMade: 0,
+          expiry: validDate,
+          value: 456
+        }
+      };
+      usersService.findByEmail.mockResolvedValue(user);
+      const result = await authService.sendVerificationCode(currentUser, locale);
+      // check to see if the logic inside sendVerificationCode works with a valid date
+      expect(result.attemptsMade).toBe(0);
+      expect(result.expiry).toBe(validDate);
+    });
+
+    it('should update the user.verificationCode when the date is expired', async () => {
+      // Use Record<string, any> to have updateOne function within the User type
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const user: User & Record<string, any> = {
```

Comment:
for next time, extract this kind of largeish object out to a stub, or use `beforeEach`. It is fine for the purposes of this though.